<div class="alert alert-block alert-info" style="background-color: #A0C020; color: white;">
<center> <h1> Business Case 1 - Hotel Customer Segmentation </h1> </center> <br>
<center> <h2> Group P </h2> </center> <br>
    <center> <h3> 2024/2025 </h3> </center>

# BUSINESS CASE 1: HOTEL CUSTOMER SEGMENTATION

### This notebook uses the *Case1_HotelCustomerSegmentation.csv* dataset

<h3 style="text-align: center;"><u></b>Group P</u></b></h3>

|     Student Name     |     Student ID     | 
|         ---          |           ---          |
|     Beatriz Monteiro|   20240591        | 
|     Catarina Nunes           |       20230083       |
|     Margarida Raposo      |       20241020       |
|     Teresa Menezes   |      20240333       |

<div style="background-color: #A0C020; padding: 10px; border-radius: 8px; text-align: center; margin-top: 20px; margin-bottom: 20px;">
    <h2 style="color: white; margin: 0;">Project Metadata</h2>
</di>

### Project Title
**Hotel Customer Segmentation**

### Project Description
This case study focuses on customer segmentation for Hotel H, a member of an independent hotel chain located in Lisbon, Portugal. The hotel's current market segmentation strategy, based solely on the origin of customers, is deemed inadequate by the new marketing manager, A. The primary objective is to develop a more comprehensive segmentation approach that incorporates geographic, demographic, and behavioral characteristics to better understand and target different customer groups. The findings aim to inform strategic decisions on product offerings, pricing, and marketing efforts to enhance customer engagement and acquisition.

### Features
- **ID**: Customer ID
- **Nationality**: Nationality of the customer in ISO 3166-1 (Alpha 3) format
- **Age**: The age of the customer
- **DaysSinceCreation**: Number of elapsed days since the customer was created
- **NameHash**: Hash of the customer's name
- **DocIDHash**: Hash of the customer’s personal document identification number (usually a passport or ID card)
- **AverageLeadTime**: Average number of days before arrival date the customer makes bookings
- **LodgingRevenue**: Total amount of lodging revenue paid by the customer so far
- **OtherRevenue**: Total amount of other revenue (e.g., food & beverage, spa, etc.) paid by the customer so far
- **BookingsCanceled**: Number of bookings the customer made but subsequently canceled
- **BookingsNoShowed**: Number of bookings the customer made but subsequently made a "no-show"
- **BookingsCheckedin**: Number of bookings the customer made, which actually ended up staying
- **PersonNights**: Total person/nights the customer has stayed at the hotel so far. Persons/Nights are the sum of Adults and Children in 
each booking, multiplied by the number of Nights (Length-of-stay) of the booking
- **RoomNights**: Total of room/nights the customer has stayed at the hotel so far. Room/Nights are the multiplication of the number of 
rooms of each booking by the the number of Nights (Length-ofstay) of the booking
- **DistributionChannel**: Distribution channel normally used by the customer to make bookings at the hotel
- **MarketSegment**: Current market segment of the customer
- **SRHighFloor**: Indication if the customer usually asks for a room in a higher floor (0: No, 1: Yes)
- **SRLowFloor**: Indication if the customer usually asks for a room in a lower floor (0: No, 1: Yes)
- **SRAccessibleRoom**: Indication if the customer usually asks for an accessible room (0: No, 1: Yes)
- **SRMediumFloor**: Indication if the customer usually asks for a room in a middle floor (0: No, 1: Yes)
- **SRBathtub Indication**: if the customer usually asks for a room with a bathtub (0: No, 1: Yes)
- **SRShower**: Indication if the customer usually asks for a room with a shower (0: No, 1: Yes)
- **SRCrib**: Indication if the customer usually asks for a crib (0: No, 1: Yes)
- **SRKingSizeBed**: Indication if the customer usually asks for a room with a king size bed (0: No, 1: Yes)
- **SRTwinBed Indication**: if the customer usually asks for a room with a twin bed (0: No, 1: Yes)
- **SRNearElevator**: Indication if the customer usually asks for a room near the elevator (0: No, 1: Yes)
- **SRAwayFromElevator**: Indication if the customer usually asks for a room away from the elevator (0: No, 1: Yes)
- **SRNoAlcoholInMiniBar**: Indication if the customer usually asks for a room with no alcohol in the mini bar (0: No, 1: Yes)
- **SRQuietRoom**: Indication if the customer usually asks for a room away from the noise (0: No, 1: Yes)


### CRISP-DM
1. **Business Understanding**: Defining objectives, assessing resources, and project planning.
2. **Data Understanding**: Collecting, exploring, and verifying data quality.
2. **Data Preparation**: Selecting, cleaning, constructing, integrating, and formatting data to ensure it is ready for analysis.
3. **Modeling**: 
4. **Evaluation**:
5. **Deployment**:


<div style="background-color: #A0C020; padding: 10px; border-radius: 8px; text-align: center; margin-top: 20px; margin-bottom: 20px;">
    <h2 style="color: white; margin: 0;">Import Libraries</h2>
</div>

In [2]:
import pandas as pd
import numpy as np

In [3]:
pd.set_option('display.max_columns', None) # to display the max number of columns

<a id = "Index"></a>

<div style="background-color: #A0C020; padding: 10px; border-radius: 8px; text-align: center; margin-top: 20px; margin-bottom: 20px;">
    <h1 style="color: white; margin: 0;">Index</h1>
</div>

1. [Data Understanding](#Data-Understanding)
    * 1.1. [Data Loading and Description](#Data-Loading-and-Description)

    * 1.2. [Exploratory Data Analysis](#EDA)

        - [1.2.1. Data Types](#Data-Types)

        - [1.2.2. Univariate EDA: Descriptive Summary](#Descriptive-Summary)

        - [1.2.3. Missing Values](#missingvalues)

        - [1.2.4. Univariate EDA: Data Visualization](#Univariate-Vizualization)
 
        - [1.2.5. Bivariate EDA: Data Visualization](#Bivariate-Vizualization) 
 
        - [1.2.6. Multvariate EDA: Duplicates](#Multivariate) 

2. [Data Preparation](#Data-Preparation)

<div style="background-color: #A0C020; padding: 10px; border-radius: 8px; text-align: center; margin-top: 20px; margin-bottom: 20px;">
    <h1 style="color: white; margin: 0;">1. Data Understanding</h1>
</div>

<div style="background-color: #A0C020; padding: 10px; border-radius: 8px; text-align: center; margin-top: 20px; margin-bottom: 20px;">
    <h2 style="color: white; margin: 0;">1.1. Data Loading and Description</h2>
</div>

Let's start off by importing our dataset usind read_csv from pandas.

In [6]:
# Opening our dataset
df = pd.read_csv('Case1_HotelCustomerSegmentation.csv', sep=';')
df

ID Nationality   Age  DaysSinceCreation  \
0            1         PRT  52.0                440   
1            2         PRT   NaN               1385   
2            3         DEU  32.0               1385   
3            4         FRA  61.0               1385   
4            5         FRA  52.0               1385   
...        ...         ...   ...                ...   
111728  111729         DEU  31.0                 36   
111729  111730         BRA  43.0                 36   
111730  111731         BRA  37.0                 36   
111731  111732         DEU  48.0                 36   
111732  111733         DEU  48.0                 36   

                                                 NameHash  \
0       0x2C371FD6CE12936774A139FD7430C624F1C4D5109CE6...   
1       0x198CDB98BF37B6E23F9548C56A88B00912D65A9AA0D6...   
2       0xDA46E62F66936284DF2844EC4FC542D0DAD780C0EE0C...   
3       0xC45D4CD22C58FDC5FD0F95315F6EFA5A6E7149187D49...   
4       0xD2E3D5BFCA141865669F98D64CDA85AD04DEFF47F8A0...   
...                                                   ...   
111728  0x2F59F0E86596B861C3303585FDB7F090B1B2D70442C5...   
111729  0x2708B9F11C95F384129152CDF0830B566F02D42B87AC...   
111730  0xEF5F6AAA6120F2AC49179E16FD12477C243A8F4FC504...   
111731  0xC70AE03F615CB96B0C84CF29419C2DDFE8EC64B77104...   
111732  0xE52EE501843E3108F20BB8F2EC180D8654DF1424EFA2...   

                                                DocIDHash  AverageLeadTime  \
0       0x434FD3D59469C73AFEA087017FAF8CA2296493AEABDE...               59   
1       0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...               61   
2       0x27F5DF762CCDA622C752CCDA45794923BED9F1B66300...                0   
3       0x8E59572913BB9B1E6CAA12FA2C8B7BF387B1D1F3432E...               93   
4       0x42BDEE0E05A9441C94147076EDDCC47E604DA5447DD4...                0   
...                                                   ...              ...   
111728  0x86AC02B9393E7F58354BA4193D792C07DB91D4BE27C7...                0   
111729  0xE87DEB08B0D7D0BDC590949FF04AAA893018BD8EB714...              170   
111730  0x74B9B5A554F5F660371E56D4CE4C1D4C3DC80AE71D68...                0   
111731  0xF589C016E6988AECD3E3BE793E7D606BBDB6B6D4795C...               66   
111732  0xD539E89183AB389864B7A24B3C1FBC1FE20A40ED02E7...                0   

        LodgingRevenue  OtherRevenue  BookingsCanceled  BookingsNoShowed  \
0               292.00          82.3                 1                 0   
1               280.00          53.0                 0                 0   
2                 0.00           0.0                 0                 0   
3               240.00          60.0                 0                 0   
4                 0.00           0.0                 0                 0   
...                ...           ...               ...               ...   
111728            0.00           0.0                 0                 0   
111729          755.25          20.0                 0                 0   
111730            0.00           0.0                 0                 0   
111731          708.00         185.0                 0                 0   
111732            0.00           0.0                 0                 0   

        BookingsCheckedIn  PersonsNights  RoomNights    DistributionChannel  \
0                       2              6           4              Corporate   
1                       1             10           5  Travel Agent/Operator   
2                       0              0           0  Travel Agent/Operator   
3                       1             10           5  Travel Agent/Operator   
4                       0              0           0  Travel Agent/Operator   
...                   ...            ...         ...                    ...   
111728                  0              0           0  Travel Agent/Operator   
111729                  1             10           5  Travel Agent/Operator   
111730                  0              0           0  Tra

Now, we want to set the column `ID` as the index in the dataframe, as this column is supposed to identify an individual reservation. However, first, to ensure we don't lose any important rows due to identifier mistakes, we will check if this column only contains unique values:

In [7]:
unique_reservations = df['ID'].nunique()
total_rows = len(df)

print(f"The ID column has {unique_reservations}"
      f" unique values,\nand the dataframe df has {total_rows} rows.")
if unique_reservations == total_rows:
    print("All ID values are unique.")
else:
    print("There are duplicate ID values.")

The ID column has 111733 unique values,
and the dataframe df has 111733 rows.
All ID values are unique.


In [8]:
# Set ID as the index
df.set_index('ID', inplace=True)

In [9]:
df.head(5)

Nationality   Age  DaysSinceCreation  \
ID                                        
1          PRT  52.0                440   
2          PRT   NaN               1385   
3          DEU  32.0               1385   
4          FRA  61.0               1385   
5          FRA  52.0               1385   

                                             NameHash  \
ID                                                      
1   0x2C371FD6CE12936774A139FD7430C624F1C4D5109CE6...   
2   0x198CDB98BF37B6E23F9548C56A88B00912D65A9AA0D6...   
3   0xDA46E62F66936284DF2844EC4FC542D0DAD780C0EE0C...   
4   0xC45D4CD22C58FDC5FD0F95315F6EFA5A6E7149187D49...   
5   0xD2E3D5BFCA141865669F98D64CDA85AD04DEFF47F8A0...   

                                            DocIDHash  AverageLeadTime  \
ID                                                                       
1   0x434FD3D59469C73AFEA087017FAF8CA2296493AEABDE...               59   
2   0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...               61   
3   0x27F5DF762CCDA622C752CCDA45794923BED9F1B66300...                0   
4   0x8E59572913BB9B1E6CAA12FA2C8B7BF387B1D1F3432E...               93   
5   0x42BDEE0E05A9441C94147076EDDCC47E604DA5447DD4...                0   

    LodgingRevenue  OtherRevenue  BookingsCanceled  BookingsNoShowed  \
ID                                                                     
1            292.0          82.3                 1                 0   
2            280.0          53.0                 0                 0   
3              0.0           0.0                 0                 0   
4            240.0          60.0                 0                 0   
5              0.0           0.0                 0                 0   

    BookingsCheckedIn  PersonsNights  RoomNights    DistributionChannel  \
ID                                                                        
1                   2              6           4              Corporate   
2                   1             10           5  Travel Agent/Operator   
3                   0              0           0  Travel Agent/Operator   
4                   1             10           5  Travel Agent/Operator   
5                   0              0           0  Travel Agent/Operator   

            MarketSegment  SRHighFloor  SRLowFloor  SRAccessibleRoom  \
ID                                                                     
1               Corporate            0           0                 0   
2   Travel Agent/Operator            0           0                 0   
3   Travel Agent/Operator            0           0                 0   
4   Travel Agent/Operator            0           0                 0   
5   Travel Agent/Operator            0           0                 0   

    SRMediumFloor  SRBathtub  SRShower  SRCrib  SRKingSizeBed  SRTwinBed  \
ID                                                                         
1               0          0         0       0              0          0   
2               0          0         0       0              0          0   
3               0          0         0       0              0          0   
4               0          0         0       0              0          0   
5               0          0         0       0              0          0   

    SRNearElevator  SRAwayFromElevator  SRNoAlcoholInMiniBar  SRQuietRoom  
ID                                                                         
1                0                   0                     0            0  
2                0                   0                     0            0  
3                0                   0                     0            0  
4                0                   0                     0            0  
5                0                   0                     0            0

In [10]:
# We create a copy to not cause any issues on main dataframe.
df_copy = df.copy()

<a id = "EDA"></a>

<div style="background-color: #A0C020; padding: 10px; border-radius: 8px; text-align: center; margin-top: 20px; margin-bottom: 20px;">
    <h2 style="color: white; margin: 0;">1.2. Exploratory Data Analysis</h2>
</div>

<a class="anchor" id="Data-Types">

### 1.2.1. Data Types

</a>

In [11]:
df.dtypes

Nationality              object
Age                     float64
DaysSinceCreation         int64
NameHash                 object
DocIDHash                object
AverageLeadTime           int64
LodgingRevenue          float64
OtherRevenue            float64
BookingsCanceled          int64
BookingsNoShowed          int64
BookingsCheckedIn         int64
PersonsNights             int64
RoomNights                int64
DistributionChannel      object
MarketSegment            object
SRHighFloor               int64
SRLowFloor                int64
SRAccessibleRoom          int64
SRMediumFloor             int64
SRBathtub                 int64
SRShower                  int64
SRCrib                    int64
SRKingSizeBed             int64
SRTwinBed                 int64
SRNearElevator            int64
SRAwayFromElevator        int64
SRNoAlcoholInMiniBar      int64
SRQuietRoom               int64
dtype: object

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 111733 entries, 1 to 111733
Data columns (total 28 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   Nationality           111733 non-null  object 
 1   Age                   107561 non-null  float64
 2   DaysSinceCreation     111733 non-null  int64  
 3   NameHash              111733 non-null  object 
 4   DocIDHash             110732 non-null  object 
 5   AverageLeadTime       111733 non-null  int64  
 6   LodgingRevenue        111733 non-null  float64
 7   OtherRevenue          111733 non-null  float64
 8   BookingsCanceled      111733 non-null  int64  
 9   BookingsNoShowed      111733 non-null  int64  
 10  BookingsCheckedIn     111733 non-null  int64  
 11  PersonsNights         111733 non-null  int64  
 12  RoomNights            111733 non-null  int64  
 13  DistributionChannel   111733 non-null  object 
 14  MarketSegment         111733 non-null  object 
 15  SRHig

`Note`: Its important to optimize the columns datatyes, specially in large models to optimize the model performance. Right now we have a memory usage of 24.7+ MB, but we can clearly see that some columns can be optimized. 

- Lets see the values of each column to see the optimal type for the Categorical Variables Columns and Numerical Variables: 

<a class="anchor" id="Descriptive-Summary">

### 1.2.2. Univariate EDA: Descriptive Summary

</a>

In [13]:
# Summary statistics for numeric columns
df.describe().T

count        mean         std   min    25%    50%  \
Age                   107561.0   45.639191   17.244952 -10.0   33.0   47.0   
DaysSinceCreation     111733.0  595.026599  374.657382  36.0  288.0  522.0   
AverageLeadTime       111733.0   60.833147   85.115320  -1.0    0.0   21.0   
LodgingRevenue        111733.0  283.851283  379.131556   0.0    0.0  208.0   
OtherRevenue          111733.0   64.682802  123.580715   0.0    0.0   31.0   
BookingsCanceled      111733.0    0.002282    0.080631   0.0    0.0    0.0   
BookingsNoShowed      111733.0    0.000600    0.028217   0.0    0.0    0.0   
BookingsCheckedIn     111733.0    0.737607    0.730889   0.0    0.0    1.0   
PersonsNights         111733.0    4.328318    4.630739   0.0    0.0    4.0   
RoomNights            111733.0    2.203825    2.301637   0.0    0.0    2.0   
SRHighFloor           111733.0    0.042512    0.201755   0.0    0.0    0.0   
SRLowFloor            111733.0    0.001307    0.036125   0.0    0.0    0.0   
SRAccessibleRoom      111733.0    0.000224    0.014957   0.0    0.0    0.0   
SRMediumFloor         111733.0    0.000770    0.027733   0.0    0.0    0.0   
SRBathtub             111733.0    0.003132    0.055881   0.0    0.0    0.0   
SRShower              111733.0    0.001629    0.040327   0.0    0.0    0.0   
SRCrib                111733.0    0.016181    0.126173   0.0    0.0    0.0   
SRKingSizeBed         111733.0    0.363268    0.480943   0.0    0.0    0.0   
SRTwinBed             111733.0    0.156811    0.363624   0.0    0.0    0.0   
SRNearElevator        111733.0    0.000331    0.018195   0.0    0.0    0.0   
SRAwayFromElevator    111733.0    0.003598    0.059874   0.0    0.0    0.0   
SRNoAlcoholInMiniBar  111733.0    0.000197    0.014031   0.0    0.0    0.0   
SRQuietRoom           111733.0    0.087718    0.282886   0.0    0.0    0.0   

                        75%       max  
Age                    58.0    123.00  
DaysSinceCreation     889.0   1385.00  
AverageLeadTime        95.0    588.00  
LodgingRevenue        393.3  21781.00  
OtherRevenue           84.0   8859.25  
BookingsCanceled        0.0     15.00  
BookingsNoShowed        0.0      3.00  
BookingsCheckedIn       1.0     76.00  
PersonsNights           6.0    116.00  
RoomNights              3.0    185.00  
SRHighFloor             0.0      1.00  
SRLowFloor              0.0      1.00  
SRAccessibleRoom        0.0      1.00  
SRMediumFloor           0.0      1.00  
SRBathtub               0.0      1.00  
SRShower                0.0      1.00  
SRCrib                  0.0      1.00  
SRKingSizeBed           1.0      1.00  
SRTwinBed               0.0      1.00  
SRNearElevator          0.0      1.00  
SRAwayFromElevator      0.0      1.00  
SRNoAlcoholInMiniBar    0.0      1.00  
SRQuietRoom             0.0      1.00

As mentioned before, we can see that some datatypes are inefficient.

**Categorical Variables**: `Nationality`; `NameHash`; `DocIDHash`; `DistributionChannel`; `MarketSegment` are correctly defined as object.

**Numeric variables**: `Age`; `DaysSinceCreation`; `AverageLeadTime`; `LodgingRevenue`; `OtherRevenue`; `BookingsCanceled`; `BookingsNoShowed`; `BookingsCheckedIn`; `PersonsNights`; `RoomNights`.
- The only numeric columns that should be in float format are `LodgingRevenue` and `OtherRevenue`. Large range (0 to 21,781) and (0 to 8859.25), so **float32** is optimal for memory without losing precision.
- The variable `age` contains incorrect negative values, the type should be **uint8** because no one is older than 255 years and there are no negative ages.
- `DaysSinceCreation` ranges from 36 to 1385, fits within uint16 (0 to 65,535).
- `AverageLeadTime` has a negative value of -1 that should be an error because doesn't make sense to have negative values in a variable that represents the *average number of days before arrival date 
the customer makes bookings*. So the type should be **int16** if we want to consider the negative value or **uint16** (0 to 65,535) if not.
- For variables `BookingsCanceled`, `BookingsNoShowed`, `BookingsCheckedIn` and `PersonsNights`the type **uint8** (0 to 255) is the best choice.

**Boolean (0: No, 1: Yes)**: `SRHighFloor`; `SRLowFloor`; `SRAccessibleRoom`; `SRMediumFloor`; `SRBathtub`; `SRShower`; `SRCrib`; `SRKingSizeBed`; `SRTwinBed`; `SRNearElevator`; `SRAwayFromElevator`; `SRNoAlcoholInMiniBar`; `SRQuietRoom`.
- Boolean variables should be of type **`bool`** or **`int8`** to optimize memory usage.

| column                                   | corrected type      |
|------------------------------------------|----------------------|
| `Age`                                    | `uint8`              | 
| `DaysSinceCreation`                      | `uint16`            | 
| `AverageLeadTime`                        | `uint16`           |
| `LodgingRevenue`                         | `float32`              | 
| `OtherRevenue`                           | `float32`              | 
| `BookingsCanceled`                       | `uint8`              | 
| `BookingsNoShowed`                       | `uint8`              | 
| `BookingsCheckedIn`                      | `uint8`              | 
| `APersonsNights`                         | `uint8`              | 
| `SRHighFloor`; `SRLowFloor`; `SRAccessibleRoom`; `SRMediumFloor`; `SRBathtub`; `SRShower`; `SRCrib`; `SRKingSizeBed`; `SRTwinBed`; `SRNearElevator`; `SRAwayFromElevator`; `SRNoAlcoholInMiniBar`; `SRQuietRoom`                   | `uint8` or `bool`             | 

In [15]:
# Summary statistics for non numeric columns
df.describe(include = ['O']).T

count  unique  \
Nationality          111733     199   
NameHash             111733  107584   
DocIDHash            110732  103480   
DistributionChannel  111733       4   
MarketSegment        111733       7   

                                                                   top   freq  
Nationality                                                        FRA  16516  
NameHash             0x15A713CE687991691A18F6CDC56ABE24979C73CF5D51...     75  
DocIDHash            0xE3B0C44298FC1C149AFBF4C8996FB92427AE41E4649B...   3032  
DistributionChannel                              Travel Agent/Operator  91019  
MarketSegment                                                    Other  63680

<a class="anchor" id="missingvalues">

### 1.2.3. Missing Values

</a>

In [16]:
# Checking missing values - Code adapted from prev notebook
missing_values = df.isnull().sum().to_frame(name='Total')
missing_values['Percentage'] = (missing_values['Total'] / len(df) * 100).round(2).astype(str) + '%'
missing_values.sort_values(by='Total', ascending=False, inplace=True)
missing_values

Total Percentage
Age                    4172      3.73%
DocIDHash              1001       0.9%
Nationality               0       0.0%
SRHighFloor               0       0.0%
SRNoAlcoholInMiniBar      0       0.0%
SRAwayFromElevator        0       0.0%
SRNearElevator            0       0.0%
SRTwinBed                 0       0.0%
SRKingSizeBed             0       0.0%
SRCrib                    0       0.0%
SRShower                  0       0.0%
SRBathtub                 0       0.0%
SRMediumFloor             0       0.0%
SRAccessibleRoom          0       0.0%
SRLowFloor                0       0.0%
MarketSegment             0       0.0%
DistributionChannel       0       0.0%
RoomNights                0       0.0%
PersonsNights             0       0.0%
BookingsCheckedIn         0       0.0%
BookingsNoShowed          0       0.0%
BookingsCanceled          0       0.0%
OtherRevenue              0       0.0%
LodgingRevenue            0       0.0%
AverageLeadTime           0       0.0%
NameHash                  0       0.0%
DaysSinceCreation         0       0.0%
SRQuietRoom               0       0.0%

Only columns `Age` and `DocIDHash` have missing values, 3.73% and 0.9% respectively. Different methods can be used to correct these columns during preparation. 